### PR TITLE
fix: add component usage tracking, closes #487

### LINKED ISSUE #487: Implement Gearbox components usage logging
"decorate" each call with a `setOneTimeSdkHeader()`

### PR DESCRIPTION
**Pull request by polomani**
made 2 examples:
- assetTree (class component)
- timeseriesPreview (functional component with hooks)

what's good: 
- nice support for functional components

issues: 
- requests from cached context are not being tracked

looks bad:
- class components (a bit tricky to implement, but maybe we'll get rid of them later anyway?)

### CODE REVIEW

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-505.gearboxjs.preview.cogniteapp.com

File: src/components/ClientSDKProvider/ClientSDKProvider.tsx [hunk 1]
```diff
@@ -13,6 +15,10 @@ export const ClientSDKProvider: React.FC<ClientSDKProviderProps> = ({
   children,
 }) => (
   <ClientSDKContext.Provider value={client}>
-    <ClientSDKCacheProvider client={client}>{children}</ClientSDKCacheProvider>
+    <ClientSDKCacheProvider client={client}>
+      <ClientSDKProxiedContext.Provider value={wrapInProxy(client)}>
```
Comment (mrtraser) [inline src/components/ClientSDKProvider/ClientSDKProvider.tsx]: Can we have it as a separate component (like `ClientSDKCacheProvider`) to apply all proxy logic there?

File: src/context/proxiedCogniteClient.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            return originalProperty(...args);
+          };
+        } else {
+          return originalProperty;
+        }
+      },
+    };
+  }
+
+  return function forComponent(component: string) {
```
Comment (mrtraser) [inline src/context/proxiedCogniteClient.ts]: I would use arrow function here

File: src/context/proxiedCogniteClient.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    get(target, name: ClientApiName) {
+      if (apiNames.includes(name)) {
+        return new Proxy(target[name], createApiHandler());
+      } else {
+        return target[name];
+      }
+    },
+  };
+
+  function createApiHandler<T extends ClientApi>(): ProxyHandler<T> {
```
Comment (mrtraser) [inline src/context/proxiedCogniteClient.ts]: I would replace this with (arrow function):
```typescript
  const createApiHandler = <T extends ClientApi>(): ProxyHandler<T> => {
    return {
      get(target: T, propKey: keyof T) {
        const originalProperty = target[propKey];
        if (typeof originalProperty === 'function') {
          return <Args>(...args: Args[]) => {
            client!.setOneTimeSdkHeader(getComponentHeader(componentName));
            return originalProperty(...args);
          };
        } else {
          return originalProperty;
        }
      },
    };
  };
```

File: src/context/proxiedCogniteClient.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+type ClientApiName = typeof apiNames[number];
+type ClientApi = CogniteClient[ClientApiName];
+
+const GearboxHeader = `CogniteGearbox:${version}`;
+
+function getComponentHeader(component: string) {
+  return `${GearboxHeader}/${component}`;
+}
+
+export function wrapInProxy(client: CogniteClient | null) {
```
Comment (mrtraser) [inline src/context/proxiedCogniteClient.ts]: add type here `ClientSDKProxiedContextType`

File: src/context/clientSDKProxiedContext.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+export type ClientSDKProxiedContextType = (
+  component: string
+) => CogniteClient | null;
+
+export const ClientSDKProxiedContext = React.createContext<
+  ClientSDKProxiedContextType
+>(() => null);
+
+export function useCogniteContext<
```
Comment (mrtraser) [inline src/context/clientSDKProxiedContext.ts]: Awesome 🚀

File: src/context/proxiedCogniteClient.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    },
+  };
+
+  function createApiHandler<T extends ClientApi>(): ProxyHandler<T> {
+    return {
+      get(target: T, propKey: keyof T) {
+        const originalProperty = target[propKey];
+        if (typeof originalProperty === 'function') {
+          return <Args>(...args: Args[]) => {
+            client!.setOneTimeSdkHeader(getComponentHeader(componentName));
```
Comment (fredrik955) [inline src/context/proxiedCogniteClient.ts]: What if the user of gearbox is using an old version of the SDK?

Comment (polomani) [reply src/context/proxiedCogniteClient.ts]: I think we updated our peer dependency, but it's a good point

Comment (polomani) [reply src/context/proxiedCogniteClient.ts]: made a check for that

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-505.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-505.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-505.gearboxjs.preview.cogniteapp.com

**Comment by cognite-cicd** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-505.gearboxjs.preview.cogniteapp.com

**Pull request merged**